test/system/src/test/java/io/pravega/test/system/StreamCutsTest.java

```diff
@@ -0,0 +1,279 @@
+                StreamConfiguration.builder().scope(SCOPE).streamName(STREAM_ONE)
+                                   .scalingPolicy(ScalingPolicy.fixed(PARALLELISM_ONE)).build()));
+        assertTrue("Creating stream two", streamManager.createStream(SCOPE, STREAM_TWO,
+                StreamConfiguration.builder().scope(SCOPE).streamName(STREAM_TWO)
+                                   .scalingPolicy(ScalingPolicy.fixed(PARALLELISM_TWO)).build()));
+    }
+
+    /**
+     * This test verifies the correct operation of readers using StreamCuts. Concretely, the test creates two streams
+     * with different number of segments and it writes some events (TOTAL_EVENTS) in them. Afterwards, the test creates
+     * a list of StreamCuts that encompasses both streams every CUT_SIZE events. The test asserts that new groups of
+     * readers can be initialized at these StreamCut intervals and that only CUT_SIZE events are read. Finally, this
+     * test also verifies that an existing reader group can be reset to the starting position in the stream.
+     */
+    @Test
+    public void streamCutsTest() {
+        @Cleanup
+        ClientFactory clientFactory = ClientFactory.withScope(SCOPE, controllerURI);
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(SCOPE, controllerURI);
+        readerGroupManager.createReaderGroup(READER_GROUP, ReaderGroupConfig.builder().stream(Stream.of(SCOPE, STREAM_ONE))
+                                                                            .stream(Stream.of(SCOPE, STREAM_TWO))
+                                                                            .build());
+        @Cleanup
+        ReaderGroup readerGroup = readerGroupManager.getReaderGroup(READER_GROUP);
+
+        // First, write half of events in each Stream.
+        writeDummyEvents(clientFactory, STREAM_ONE, TOTAL_EVENTS / 2);
+        writeDummyEvents(clientFactory, STREAM_TWO, TOTAL_EVENTS / 2);
+        log.debug("Finished writing events to streams.");
+
+        // Second, get StreamCuts for each slice from both Streams at the same time (may be different in each execution).
+        List<Map<Stream, StreamCut>> streamSlices = getStreamCutSlices(clientFactory, readerGroup);
+        log.debug("Finished creating StreamCuts.");
+
+        // Third, ensure that reader groups can correctly read slices from different Streams.
+        int readEvents, groupId = 0;
+        Map<Stream, StreamCut> startingPoint = null;
+        final int parallelSegments = PARALLELISM_ONE + PARALLELISM_TWO;
+        ReaderGroupConfig.ReaderGroupConfigBuilder configBuilder = ReaderGroupConfig.builder()
+                                                                                    .stream(Stream.of(SCOPE, STREAM_ONE))
+                                                                                    .stream(Stream.of(SCOPE, STREAM_TWO));
+        for (Map<Stream, StreamCut> endingPoint : streamSlices) {
```

Comment:
This is good, but I'm wondering if we can have a second test case in which we shuffle the cuts such that we don't simply follow the order in which they have been generated. For example, say that we generate 4 cuts in this order: C1, C2, C3, C4. There are various combinations we can try:

- [C1, C2), [C1, C3), [C1, C4)
- [C2, C3), [C2, C4)
- [C3, C4)

We may want to test all of those.

Reply:
I added a new method that performs these combinations and reads the appropriate stream slices.